mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala

```diff
@@ -135,25 +135,39 @@ class GaussianMixture private (
     val breezeData = data.map(_.toBreeze).cache()
     
     // Get length of the input vectors
-    val d = breezeData.first().length
-    
+    val tmp = breezeData.first()
+    val (d, nFeatures) = (tmp.length, tmp.size)
+
+    val distributeGaussian = if (k >= 10 && nFeatures >= 10) true else false
     // Determine initial weights and corresponding Gaussians.
     // If the user supplied an initial GMM, we use those values, otherwise
     // we start with uniform weights, a random mean from the data, and
     // diagonal covariance matrices using component variances
-    // derived from the samples    
+    // derived from the samples.
     val (weights, gaussians) = initialModel match {
       case Some(gmm) => (gmm.weights, gmm.gaussians)
       
       case None => {
         val samples = breezeData.takeSample(withReplacement = true, k * nSamples, seed)
-        (Array.fill(k)(1.0 / k), Array.tabulate(k) { i => 
-          val slice = samples.view(i * nSamples, (i + 1) * nSamples)
-          new MultivariateGaussian(vectorMean(slice), initCovariance(slice)) 
-        })
+        val weights = Array.fill(k)(1.0 / k)
+        val gaussians = {
+          if (distributeGaussian) {
```

Comment:
Just a second, please do not review yet.  I need to make a few other changes as well.